File: JS/Dynamic Programming/1043. Partition Array for Maximum Sum.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */
var maxSumAfterPartitioning = function(arr, k) {  // DP, dp[i] = dp[i - j] + j * Max(i - 1 ... i - j)
    const n = arr.length;
    const dp = [0];
    for(let i = 1; i <= n; i++) {
        let max = -Infinity;
        let best = 0;
        for(let j = 1; j <= k && i - j >= 0; j++) {
            max = Math.max(max, arr[i - j]);
            best = Math.max(best, dp[i - j] + max * j);
        }
        dp[i] = best;
    }
    return dp[n];
};

File: Golang/DynamicProgramming/1043.partition-array-for-maximum-sum.go
package main

func maxSumAfterPartitioning(arr []int, k int) int {
	n := len(arr)
	dp := make([]int, n+1)
	min := func(a, b int) int {
		if a < b {
			return a
		}
		return b
	}
	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	for i := 1; i <= n; i++ {
		temp := -1
		for j := 1; j <= min(i, k); j++ {
			temp = max(temp, arr[i-j])
			dp[i] = max(dp[i], dp[i-j]+temp*j)
		}
	}
	return dp[n]
}
